xla/client/xla_builder.cc

```diff
@@ -2883,10 +2883,30 @@ XlaOp XlaBuilder::AllGatherImpl(const XlaOp operand,
     HloInstructionProto instr;
     TF_ASSIGN_OR_RETURN(const Shape* operand_shape, GetShapePtr(operand));
 
-    TF_ASSIGN_OR_RETURN(
-        Shape inferred_shape,
-        ShapeInference::InferAllGatherShape({operand_shape},
-                                            all_gather_dimension, shard_count));
+    std::vector<const Shape*> operand_shapes;
+    std::vector<XlaOp> operands;
+    if (operand_shape->IsTuple()) {
+      if (operand_shape->tuple_shapes_size() == 0) {
+        return Unimplemented("0 element tuple AllGather is not supported");
+      }
+      for (int i = 0; i < operand_shape->tuple_shapes_size(); ++i) {
+        if (operand_shape->tuple_shapes(i).element_type() !=
```

Comment:
why is this required here? I don't think an all-gather in particular has this restriction at HLO level. For all-reduce such restriction makes sense since the reduction computation is typed.

Also, can we share the tuple flattening code with all-reduce/reduce-scatter and make a helper function for that?

Reply:
@jurahul you are right. Removed the check for same shape.

The other request of sharing tuple flattening code will be a larger refactor across existing  all-reduce/reduce-scatter so I want to avoid that to keep the changes to minimum and self-contained.